### PR TITLE
Allow multiple solvers to find a root in HVAC system simulations

### PR DESCRIPTION
**Pull request by lgu1234**
Pull request overview
---------------------
Please change this line to a description of the pull request, with useful supporting information including whether it is a new feature, or fixes a defect, a cross reference to any defects addressed in this PR, the type of changes to be made, and whether diffs are expected/justified based on this change.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by Myoldmopar** [comment]
@lgu1234 I do have some thoughts on this NFP.  I do realize I am coming in a bit late since you already sent it out for comment, so I will put my thoughts here and you can do with them what you will.
 - This overall seems like a great idea.  Making the core root solver in E+ more intelligent makes great sense.  My instinct is that using the bisection to make a high level pass feels more reasonable, but I will defer to @rraustad as he is more intimately familiar with the solution behavior of the HVAC systems.
 - As for the input object.  I don't really like that.  If this is most useful to allow developers to evaluate/play with the solver behavior, then I don't think it should be in the IDD at all.  It is possible we could include it as a hidden input object that's not even in the IDD.  But my opinion is that developer-only options should be more likely included as environment variables.

Let me know if these seem reasonable or if you have any other questions about them.

**Comment by lgu1234** [comment]
@Myoldmopar 

Edwin:

Thanks for your comments.

Here are my thoughts regarding your comments.

This new feature proposed two issues:

1.	Create a new object to provide algorithm choices for advanced users. This is a global change. Although I can hard-wire code as environments variables. It loses flexibility for users to make choices from inputs only. In addition, hard-wire code is based on my selection. Users may have more thoughts than mine. That is why I propose a new object.
2.     Add optional arguments in existing functions for developers for local changes. The change will eliminate TightenWaterFlowLimits and others. I did unit tests and found it works. It is better than a step search with a fixed increment.

Since there are more than 3 weeks after submission, I hope to wrap it this week.

Please let me know what you think.

File: src/EnergyPlus/General.cc
Comment (rraustad) [inline src/EnergyPlus/General.cc]: I wonder if a Select Case would be faster here. There would be 5 cases, but only 1 IF ELSE statement in three of them.

Comment (lgu1234) [reply src/EnergyPlus/General.cc]: @rraustad Thanks for comments. I am open to anything. Jason told me that he prefers to use if statements, compared to use Select Case. @Myoldmopar @mjwitte Please let me know your opinions.

Comment (Myoldmopar) [reply src/EnergyPlus/General.cc]: I prefer switch, but it's not necessarily worth holding up the entire PR for that change alone.  If there is more work to do then I'd suggest making the change.  If there isn't, then don't worry much about it.

Comment (jasondegraw) [reply src/EnergyPlus/General.cc]: @lgu1234 @Myoldmopar I'd prefer a `switch` here too, but it won't work because `case` statements can't have a `const int` in them. To go that direction, I think we'd want `Bisection` etc. to be in an enum:

```
enum class HVACSystemRootFindingAlgorithm { RegulaFalsi, Bisection, 
    RegulaFalsiThenBisection, BisectionThenRegulaFalsi, Alternation };
```

and then the code can look like

```
switch( HVACSystemRootFinding.Type ) {
case HVACSystemRootFindingAlgorithm::Bisection:
  // do stuff
default:
  //do other stuff
}
```

We've talked about this in the past, and it is something we need to start doing. I don't know if now is the time, previously output issues held it up. My vote is "yes", since this is a new object. But I will leave it to you guys.

Comment (mjwitte) [reply src/EnergyPlus/General.cc]: I would say change it now.

Comment (lgu1234) [reply src/EnergyPlus/General.cc]: OK. I will make the change. Before I do it. Here is what Rich suggested:

        switch ( HVACSystemRootFinding.TypeNum )

        {
            case RegulaFalsi : {
                XTemp = ( Y0 * X1 - Y1 * X0 ) / DY;
            }
            case Bisection : {
                XTemp = ( X1 + X0 ) / 2.0;
                break;
            }
            case Alternation : {
                if ( AltIte > HVACSystemRootFinding.NumOfIter ) {
                   if ( AltIte >= 2 * HVACSystemRootFinding.NumOfIter ) AltIte = 0;
                   XTemp = ( X1 + X0 ) / 2.0;
                } else {
                   XTemp = ( Y0 * X1 - Y1 * X0 ) / DY;
                }
            }
            case ...
        }

@jasondegraw Do I have to use enum class for Switch statements?

Comment (rraustad) [reply src/EnergyPlus/General.cc]: Also, how does the `break` work here? If this does what I hope, then the break should not be necessary, correct?

Comment (lgu1234) [reply src/EnergyPlus/General.cc]: @rraustad Here is a section extracted from MSDN:

// switch_statement1.cpp  
#include <stdio.h>  
  
int main() {  
   char *buffer = "Any character stream";  
   int capa, lettera, nota;  
   char c;  
   capa = lettera = nota = 0;  
  
   while ( c = *buffer++ )   // Walks buffer until NULL  
   {  
      switch ( c )  
      {  
         case 'A':  
            capa++;  
            break;  
         case 'a':  
            lettera++;  
            break;  
         default:  
            nota++;  
      }  
   }  
   printf_s( "\nUppercase a: %d\nLowercase a: %d\nTotal: %d\n",  
      capa, lettera, (capa + lettera + nota) );  
}  

Break is necessary unless with case Default.

Comment (jasondegraw) [reply src/EnergyPlus/General.cc]: @lgu1234 An `enum` is not required, but it does make things more clear (in my opinion).

If there's fall-through in the switch, it'd best to include a comment so that it is clear that is the intent. Some compilers/IDEs/checkers are likely to complain.

**Comment by Myoldmopar** [comment]
@lgu1234 Let me know when you are done working on this for a bit because I want to remove the "e" prefix from your new enum class members as they are just kinda unnecessary.  I'm happy to make the change just let me know when you are at a down point so I can make my changes, then you can pull, and we'll be able to avoid conflicts.

**Comment by lgu1234** [comment]
@Myoldmopar I am done. Please feel free to make any changes.

**Comment by Myoldmopar** [comment]
Great.  I'll push momentarily.  Just make sure you pull before making any more changes.

**Comment by nrel-bot-2** [comment]
@lgu1234 @lgentile it has been 16 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
@lgu1234 I resolved the variable name in the unit test issue, so this should build fine again.

However, there is a merge conflict.  It looks like the solution is likely to just keep a chunk of code in this branch, but I couldn't tell and didn't want to cause problems by improperly resolving the conflict.

If you could pull this branch, then try merging the latest develop in, you'll see the conflict in `WaterCoils.cc`.  Let me know if you need to chat about resolving the conflict.

File: idd/Energy+.idd.in
Comment (Myoldmopar) [inline idd/Energy+.idd.in]: Tab character here; I can clean it up.

File: idd/Energy+.idd.in
Comment (Myoldmopar) [inline idd/Energy+.idd.in]: "Number of Iteration", or
"Number of Iterations" ?

**Review by Myoldmopar — CHANGES_REQUESTED** [review]
@lgu1234 I noticed a tab character in the IDD, I can take care of that.  I also had a question about the second field name.

@mjwitte Do you have anything further?  I'm going to re-run the new example file to peruse the error file and spot check things, but I think this is otherwise ready to go.

File: idd/Energy+.idd.in
Comment (lgu1234) [reply idd/Energy+.idd.in]: @Myoldmopar I think it should be "Number of Iterations". I am going to make changes.

File: idd/Energy+.idd.in
Comment (lgu1234) [reply idd/Energy+.idd.in]: @Myoldmopar I will clean up, since I am gong to make changes based your another comments.

**Comment by lgu1234** [comment]
@Myoldmopar Comments are addressed. The files are uploaded.

File: idd/Energy+.idd.in
Comment (mjwitte) [inline idd/Energy+.idd.in]: @lgu1234 Just one comment here.  Minor detail, but I would place this object after Timestep or after ConvergenceLimits.  But maybe it doesn't matter.  The whole order of things in this section of the IDD doesn't make much sense anymore.
@Myoldmopar This one's all yours.  I have not further comments.

Comment (lgu1234) [reply idd/Energy+.idd.in]: @mjwitte I am going to move the object after ConvergenceLimits.

**Comment by Myoldmopar** [comment]
@lgu1234 Would it be a quick fix to clean up the error file on the new example file?  It has:

```
   ** Warning ** The Standard Ratings is calculated for Coil:Cooling:DX:MultiSpeed = HEAT PUMP ACDXCOIL 1 but not at the AHRI test condition due to curve out of bound.
   **   ~~~   **  Review the Standard Ratings calculations in the Engineering Reference for this coil type. Also, use Output:Diagnostics, DisplayExtraWarnings for further guidance.
   ** Warning ** The Standard Ratings is calculated for Coil:Cooling:DX:MultiSpeed = HEAT PUMP ACDXCOIL 1 but not at the AHRI test condition due to curve out of bound.
   **   ~~~   **  Review the Standard Ratings calculations in the Engineering Reference for this coil type. Also, use Output:Diagnostics, DisplayExtraWarnings for further guidance.
   ** Warning ** The Standard Ratings is calculated for Coil:Cooling:DX:MultiSpeed = HEAT PUMP ACDXCOIL 1 but not at the AHRI test condition due to curve out of bound.
   **   ~~~   **  Review the Standard Ratings calculations in the Engineering Reference for this coil type. Also, use Output:Diagnostics, DisplayExtraWarnings for further guidance.
   ** Warning ** The Standard Ratings is calculated for Coil:Cooling:DX:MultiSpeed = HEAT PUMP ACDXCOIL 1 but not at the AHRI test condition due to curve out of bound.
   **   ~~~   **  Review the Standard Ratings calculations in the Engineering Reference for this coil type. 
```

They're just warnings, so they aren't going to ruin anyone's day by going in, and I bet they just came over with the multi-speed file you started with.  But it's always best to have a clean error file on our examples.

I have nothing else on this branch, so let me know on that and then this can go in.

**Comment by lgu1234** [comment]
@Myoldmopar I changed the value of Minimum Value of Y from 23.88889 to 18.88880 in both curves of HPACCoolCapFT and HPACCOOLEIRFT, suggested by @Nigusse . The above warnings are gone. I am going to upload the new example file.

**Comment by Myoldmopar** [comment]
Thanks @lgu1234 I re-ran the new file and confirmed the warnings are gone.  I'm going to get this merged in now.

**Pull request merged**
